feat(0002): add numeric values (#3)
blight-js/rfcs: text/0002-numeric-values.md

```js
0 ^ 4294967296;
// JavaScript -> 0
// Blight     -> 4294967296
```

Since fixed-width integer types have a specific minimum and maximum, it is undefined behavior to error to overflow or underflow when doing arithmatic. In a safe build, this will safely crash with a stack trace and information on what values were provided, but at runtime this will cause undefined behavior.

The following operations will cause undefined behavior:

- `+`: If the result of the addition is above `Int.max`
- `-`: If the result of the subtraction is above `Int.min`.
- `*`: If the result of the multiplication is above `Int.max` or below `Int.min`.
- `/`: If there is a remainder when dividing.
- `/`: If the divisor is `0`.
- `%`: If the divisor is `0`.
- Unary `-` can underflow if when negating `Int.max`.

## Floats

There is one floating point type: `f64`, which is a 64-bit floating point that follows the [IEEE 754 standard](https://en.wikipedia.org/wiki/IEEE_754).

## Built-in Properties on the Float Type

`f64` has built in properties. These are accessed by using property access notation.

- `f64.min`: Constant `1.7976931348623157e+308`

> - `f64.max`: Constant `5e-324`

Reviewer: what...

Author: these constants are the largest and smallest values allowed for floating point, as per ieee specification

Reviewer: Are they reversed or am I reading this wrong?

Author: they are reversed. my mistake. will update the document.

- `f64.inf`: Constant `Infinity`

## Built-in Properties on Float Values

Similarly to integers, floats have built-in properties and methods. The variable name `val` is a placeholder for the float value used.

- `val.to(T)`: Converts the float's value to a the given type. T can be
  - An integer type like `u8`, which it will convert the value to that type, rounding the float if needed.
    - It is safety-checked undefined behavior if the float is out of range of `T.min`..`T.max`.

>     - It is safety-checked undefined behavior if the float is not an integer.

Reviewer: You say "rounding the float if needed" yet you have a safety check later. Which is it?

Author: will edit, it is meant to be a safety check. this is due to BigInt not automatically rounding. i will clarify this in the document.

  - `f64`, which it will do nothing.
- `val.round()`: Round this float to the nearest integer, while keeping it a float.
- `val.floor()`: Round this float down to the nearest integer, while keeping it a float.
- `val.ceil()`: Round this float up to the nearest integer, while keeping it a float.
- `val.pow(y)`: Raise this float to the power of `y`.

>   - It is safety-checked undefined behavior to call `1.pow(f64.inf)`.

Reviewer: Add a safety check for 0^0

Author: 0.pow(0) is 1. i will clarify this identity in the document.

- `/`: If the divisor is `0`.
- `%`: If the divisor is `0`.

Note that floating point operations cannot overflow, but instead will return `f64.inf` as IEEE 754 specifies.

## NaN

There is no `NaN` constant. Instead, any operation that could produce `NaN` is either safety-checked undefined behavior (like dividing by zero), or will return an optional `f64`. When representing `NaN` is required, it should be represented as an optional `f64`, to indicate the possibility of `NaN`.

If it was missed when writing this RFC. Any math operation that would produce `NaN` should be safety-checked undefined behavior.

> [!NOTE]
> There is not yet an RFC for optionals.

# Reference-level explanation

[reference-level-explanation]: #reference-level-explanation

## Parsing numbers

For reference, the following regular expression is a set of all valid numeric literals:

```regexp
^(?:\d(?:\d|_\d)*)?\.?(?:\d(?:\d|_\d)*)+(?:e[+-]?(?:\d(?:\d|_\d)*))?\b
```

A compiler should implement this in native code than using a regular expression engine.

Examples of valid numeric literals:

- `10` (an integer)
- `512_000.000` (an integer)
- `63_2.42` (a float)

> - `.425_000` (a float)

Reviewer: this one feels a bit dubious to me

Author: leading . is usually valid in most languages. _ separators are also featured in some. this test case is two of those at once.

```
var x: u32 = 512
// `512` becomes a u32 based on the type of `x`
```

Literals that do not fit in the result type will cause a compile error.

```
var x: u32 = -2
// compile error: cannot coerce `-2` to `u32`
```

You cannot use [unary `~`](#integer-arithmetic) on an integer literal, it must be given a type, such as:

```blight
~512 // compile error: cannot infer type of constant numeric literal
     // hint: unary operator '~' needs a result type

~512.to(u32) // ok
```

Similarly, runtime-mutable variables cannot be assigned without annotating their type.

> [!NOTE]
> There is not yet an RFC for variable declarations. We will assume `var name: type` is the syntax.

```blight
var x = 512 // compile error: cannot infer type of constant numeric literal

var x: u32 = 512 // ok
var x = 512.to(u32) // ok, less preferred approach
```

## Constant Expression Evaluation

Expressions that contain only numeric literals must be evaluated at compile time.

```blight

> var x: u32 = (2.5 + 2.5) / 5

Reviewer: Now I'm confused, in the compiler are these floats, bigints, or...?

Author: they are numeric literals. they do not have a type until assigned into a runtime-known variable.

// valid as `1` can coerce to `u32`

var x: u32 = 5 / 6

> // compile error: cannot coerce `0.8333333333333334` to `u32`

Reviewer: This implies that these are floats.

Author: this numeric literal being 0.8333333333333334, does not coerce into a u32.

if the literals were treated as integers, the division fails and the compile error is there, if we allow float division for constants, we get a better error message that the resulting value does not work as a valid integer.

```

If undefined behavior is caused by a constant expression, it is a compile error.

```blight
var x: u32 = 5 / 0
// compile error: division by zero
```

## Widening during arithmetic

When integers of different sizes are used in arithmetic, the compiler will widen the smaller integer to the larger integer.

For example, if a variable of type `u32` named `x` existed, and the following code was run:

```blight
x + 512
```

> `512` is widened to a u32. If the rhs was larger than `u32.max`, then this would be a compile error. Explicit widening of `x` is needed.

Reviewer: what is the rhs? The right hand side? This should be more explicit.

Author: will edit the document to clarify

```
x.to(u64) + u32.max + 123
```

## Changes to Identifiers

The following variable names are banned as they refer to built-in types, as they would shadow the implicit global.

- `/[uif][0-9]+/`

> It is [not 100% decided](#variable-naming) how this impacts identifiers in general. It should probably be the job of that proposal to decide this.

Reviewer: Why this limitation? This prevents true "first class types".

Author: this refers to declaring your own variables named after the built in ones.

for example, we assume var decl follows `const name =`, then:
```blight
const u16 = 5
```
this is now defined to be a compile error due to using the the variable name `u16` in your own declaration. i should clarify this to variable declarations.

as for property access it really isn't this rfc to dictate what identifiers are valid and what arent.

Reviewer: I think we've decided on `let` as const. Please update this section accordingly.

## Compiling `.to(T)`

As there are only number types, this function will do nothing in most cases. It may need to compile to `Number(x)` or `BigInt(x)` depending on the action taken.

## Compiling bitwise operations

Spec compliant bit-shifting is not trivial to implement, it involves preserving the size for larger integers. The simplest way to do this is to convert JS numbers into [BigInt](https://tc39.es/proposal-bigint) and use the BigInt operators. This is not ideal, as it is slower than using the JS operators directly. It is also not ideal because it is not possible to use the BigInt operators on `f64` values.

## Compiling arithmetic safety checks

> Saftey checks can be implemented somewhat trivially using a single temporary variable. This can be left as a global. It is unfortunate that so much code, as safety checks off allow this to compile to 3 characters.

Reviewer: What does this sentence mean, this isn't valid english.

Author: correction: It is unfortunate that so much code is needed to compile this with saftey checks, as with them disabled it allows this to compile to 3 characters

It is ambiguous with a potential `*` unary operator. do you parse `a**b` as:

- `a` `**` `b`
- `a` `*` `(` `*b` `)`

## Floating point overflow

Should `f64.inf + 1` be `f64.inf` or a crash. This RFC defines that it is `f64.inf`.

# Prior art

[prior-art]: #prior-art

A large part of this RFC is adapted from how Zig's [Integer](https://ziglang.org/documentation/0.11.0/#toc-Integers) and [Float](https://ziglang.org/documentation/0.11.0/#Floats) types work.

The pronounciation of the types is adapted from the C99 [stdint.h](https://en.wikipedia.org/wiki/C_data_types#stdint.h) (`uint32_t`, `int64_t`, etc).

Properties on the type definitions themselves are adapted from Rust's [Integer's Associated Constants](https://doc.rust-lang.org/std/primitive.i32.html)

# Unresolved questions

[unresolved-questions]: #unresolved-questions

## Handling Overflow Errors

A way to handle overflow errors should be given. Making these methods on the values seems like the path of least resistance, but it will lead to long code like `something.tryAdd(63)` or whatever.

Maybe later when we figure out error handling in general we can prescribe a way to handle operator errors easily.

> Similarly, this should be the case for `.sqrt`/`.pow` and so on

Reviewer: Stupid idea: what if all operations simply return an optional type if they are fallible? This could play in nicely with a future error handling proposal.

Author: either an optional type or an error type. regardless, there is no current mechanism for this, so it will be deferred until that

Reviewer: I think we should postpone error nonsense until later. I think if this spec simply says "may be an error" or something like that, that would be great.

Author: sorry, was this not under the future handling.

it was not. i meant this to be a future thing. i dont intend on solving the problem now

## Saturating and Wrapping Arithmetic

This proposal does not include saturating or wrapping arithmetic.

Two approaches to consider:

- Rust-like `left.addSaturating(right)`
- Zig-like `left +| right`

> This is a job for a different RFC, as getting the base numberic semantics working is more important.

Reviewer: Defining these terms would be nice

Author: This RFC does not define built in methods for these. It will be another one to decide this api.

Reviewer: Please move this to future possibilities.

Author: agreed

## Variable naming

Are variable names like `f64` prohibited as an identifier in general? Some languages, including JavaScript, have interesting rules for this.

Valid JavaScript:

```ts
const x = { class: 2 };
console.log(x.class);
```

Valid Zig:

```zig
pub fn main() void {
  const x = struct {
    const @"null" = 1;
    const @"u8" = 2;
  };

  const y = x.null;
  const z = x.u8;

  @import("std").debug.print("{d} {d}", .{y, z});
}
```

I'm inclined to allow it, but that isn't neccecarily about this RFC, but about a property access RFC.

## Naming compile-time known integers

This RFC does not name the type of [numeric literals](#type-of-numeric-literals). A pre-requisite to doing this realistically would be adding a full equivilent to `comptime`, which may not be desirable, and is a job for a different RFC.

## Hex/Octal/Binary literals

To keep the scope down, this RFC does not address these literals. Currently, all numeric literals are base 10.

# Future possibilities

[future-possibilities]: #future-possibilities

## More Methods

> Later on, it will be cool to add more properties / methods to the integer types, such as `i32.parse(str)`. Exactly how this would compile is not clear at the moment, as `parseInt` is a weird API that we may not want to expose as-is. The integer value can contain methods like `.string()` and others.

Reviewer: How does this differ from `i32.to(string)`?

Author: need to update this. this was in the revision before i theorized `.to(T)`

> We should also, at some point, expose ways to use the raw JS operators on `float` types.

Reviewer: 🤨

Author: there may be some instances someone would like to use some raw js operators. one instance is they may desire the raw behavior of `<<`, which will coerce the number to an i32, and then perform the bit shift, all in one motion. the other bitwise operators do similar things.

```
var x: u32 = -2
// compile error: cannot coerce `-2` to `u32`
```

You cannot use [unary `~`](#integer-arithmetic) on an integer literal, it must be given a type, such as:

```blight
~512 // compile error: cannot infer type of constant numeric literal
     // hint: unary operator '~' must operate on an integer of known width

~512.to(u32) // ok
```

Similarly, runtime-mutable variables cannot be assigned without annotating their type.

> [!NOTE]
> There is not yet an RFC for variable declarations. We will assume `var name: type` is the syntax.

```blight
var x = 512 // compile error: cannot infer type of constant numeric literal

var x: u32 = 512 // ok
var x = 512.to(u32) // ok, less preferred approach
```

## Constant Expression Evaluation

Expressions that contain only numeric literals must be evaluated at compile time.

```blight
var x: u32 = (2.5 + 2.5) / 5
// valid as `1` can coerce to `u32`
```

The implementation logic of the type resolver is as follows:

1. Calculate LHS `(2.5 + 2.5)`
   1. LHS is a numeric literal `2.5`, which is internally represented as the float `2.5`
   2. RHS is a numeric literal `2.5`, which is internally represented as the float `2.5`

>    3. Since both sides are compile-time known, this expression evaluates to the numeric literal `5`, internally represented by the float `5.0`

Reviewer: replace "float" with "f64".

```blight
x + y.to(u32)
```

Another case where this happens:

```blight
var y = x.to(u64) + u32.max + 123
```

To break down exactly how the type resolution happens here:

1. Calculate the type for the first addition:
1. The types of the operands are calculated

   - `x.to(u64)` is a `u64`
   - `u32.max` is a `u32`

1. Widening of the smaller type to match the larger type, the expression is rewritten as `x.to(u64) + u32.max.to(u64)`
1. Repeat this process on the resulting `u64` for the second addition
1. The types of the operands are calculated

   - `x.to(u64) + u32.max.to(u64)` is a `u64`
   - `123` is an numeric literal, it gets assigned a type of `u64`

1. The types are the same size, so it is okay.
1. It is rewritten such like `(x.to(u64) + u32.max.to(u64)) + 123.to(u64)`

## Changes to Identifiers

The following variable names are banned as they refer to built-in types, as they would shadow the implicit global.

- `/[uif][0-9]+/`

It is [not 100% decided](#variable-naming) how this impacts identifiers in general. It should probably be the job of that proposal to decide this.

## Compiling `.to(T)`

> As there are only number types, this function will do nothing in most cases. It may need to compile to `Number(x)` or `BigInt(x)` depending on the action taken.

Reviewer: ```suggestion
As only number types have been defined as of this RFC, this function will do nothing in most cases. It may need to compile to `Number(x)` or `BigInt(x)` depending on the action taken.
```

## Saturating and Wrapping Arithmetic

This proposal does not include saturating or wrapping arithmetic.

Two approaches to consider:

- Rust-like `left.addSaturating(right)`
- Zig-like `left +| right`

This is a job for a different RFC, as getting the base numberic semantics working is more important.

## Variable naming

Are variable names like `f64` prohibited as an identifier in general? Some languages, including JavaScript, have interesting rules for this.

Valid JavaScript:

```ts
const x = { class: 2 };
console.log(x.class);
```

Valid Zig:

```zig
pub fn main() void {
  const x = struct {
    const @"null" = 1;
    const @"u8" = 2;
  };

  const y = x.null;
  const z = x.u8;

  @import("std").debug.print("{d} {d}", .{y, z});
}
```

I'm inclined to allow it, but that isn't neccecarily about this RFC, but about a property access RFC.

## Naming compile-time known integers

This RFC does not name the type of [numeric literals](#type-of-numeric-literals). A pre-requisite to doing this realistically would be adding a full equivilent to `comptime`, which may not be desirable, and is a job for a different RFC.

## Hex/Octal/Binary literals

> To keep the scope down, this RFC does not address these literals. Currently, all numeric literals are base 10.

Reviewer: @OlexG It's mentioned in the unresolved questions section.

# Future possibilities

[future-possibilities]: #future-possibilities

## More Methods

> Later on, it will be cool to add more properties / methods to the integer types, such as `i32.parse(str)`. Exactly how this would compile is not clear at the moment, as `parseInt` is a weird API that we may not want to expose as-is. The integer value can contain methods like `.toFixed(n)` and others.

Reviewer: Still not patched. I think `i32.to(str) would be great.

> We should also, at some point, expose ways to use some of the raw JS operators on `float` types. There may be some instances someone would like to use some the JS operators directly, primarily the bit shifting ones like `^` on floating point numbers. It could be exposed on as `(5.2).lshiftInt(n)` which is defined to return an `i32`.

Reviewer: Maybe, definitely a future thing.
